### PR TITLE
Monix port

### PR DESCRIPTION
**Pull request by mariusmuja**
This is a port of Outwatch to the [Monix](https://monix.io/) library.

### CODE REVIEW

File: src/test/scala/outwatch/ScenarioTestSpec.scala [hunk 1]
```diff
@@ -193,38 +200,61 @@ class ScenarioTestSpec extends JSDomSpec {
...
+      todo <- Task {
+        val todo = "fold laundry"
+        inputElement.value = todo
+        inputElement.dispatchEvent(inputEvt)
+        submitButton.dispatchEvent(clickEvt)
+        todo
+      }
+      _ <- Task {
+        list.childElementCount shouldBe 1
+      }.executeWithFork
```
Comment (mariusmuja) [inline src/test/scala/outwatch/ScenarioTestSpec.scala]: This is because Monix uses by default a [batched execution model](https://monix.io/docs/2x/execution/scheduler.html#execution-model), so we need to "fork" the assertion to the end of the javascript execution queue, otherwise it might execute before all the effects of the event dispatch execute.

**Comment by fdietze** [comment]
Would it be worth it to port directly to a preliminary [Monix 3](https://github.com/monix/monix/releases/tag/v3.0.0-M2) Release?

**Comment by mariusmuja** [comment]
All that's needed to upgrade to Monix 3 is to update the dependency version, there seem to be no breaking changes in the API used.

**Comment by fdietze** [comment]
Interesting. Are `monix.Task` and `cats.effect.IO` compatible?

**Comment by LukaJCB** [comment]
They are!

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala [hunk 2]
```diff
@@ -143,27 +142,29 @@ private[outwatch] final case class Receivers(
...
         .groupBy(_.attribute)
         .values
-        .map(_.last.attributeStream)(breakOut)
+        .map(_.last.attributeStream)(breakOut): _*
     )
 
-    attributeReceivers.combineLatest(childStreamReceivers)
+    attributeReceivers.startWith(Seq(Seq.empty)).combineLatest(
+      childStreamReceivers.startWith(Seq(Seq.empty))
+    ).dropWhile { case (a, c) => a.isEmpty && c.isEmpty }
```
Comment (fdietze) [inline src/main/scala/outwatch/dom/helpers/DomUtils.scala]: What is this `dropWhile` good for?

File: src/main/scala/outwatch/util/Store.scala [hunk 3]
```diff
@@ -1,33 +1,37 @@
...
 import outwatch.dom._
 import outwatch.dom.helpers.STRef
-import rxscalajs.Observable
-import rxscalajs.subscription.Subscription
+
+import scala.concurrent.Future
+
+
 
 final case class Store[State, Action](initialState: State,
```
Comment (fdietze) [inline src/main/scala/outwatch/util/Store.scala]: `STRef` is only used in Store. Is it possible to remove it entirely?

File: src/main/scala/outwatch/http/Http.scala [hunk 4]
```diff
@@ -1,10 +1,40 @@
...
+import org.scalajs.dom.ext.{Ajax, AjaxException}
+import org.scalajs.dom.{Blob, XMLHttpRequest}
+
+import scala.concurrent.Future
+import scala.scalajs.js
+import scala.scalajs.js.typedarray.ArrayBuffer
+import scala.scalajs.js.|
+
 
 object Http {
```
Comment (fdietze) [inline src/main/scala/outwatch/http/Http.scala]: It seems like here is more going on than a simple port. Could you explain what you did here? I'm also not using the `Http` module myself, so this should be reviewed by someone who does.

If its too different it should be a separate commit.

File: src/test/scala/outwatch/DomEventSpec.scala [hunk 5]
```diff
@@ -418,8 +419,11 @@ class DomEventSpec extends JSDomSpec {
 
     document.getElementById("input").dispatchEvent(inputEvt)
 
-    winClicked shouldBe true
-    docClicked shouldBe true
+    Task {
```
Comment (fdietze) [inline src/test/scala/outwatch/DomEventSpec.scala]: Could you please explain, why this is neccessary here?

File: src/main/scala/outwatch/Sink.scala [hunk 6]
```diff
@@ -1,12 +1,14 @@
 package outwatch
 
 import cats.effect.IO
-import rxscalajs.facade.SubjectFacade
-import rxscalajs.subscription.Subscription
-import rxscalajs.{Observable, Observer, Subject}
+import monix.execution.Scheduler.Implicits.global
```
Comment (fdietze) [inline src/main/scala/outwatch/Sink.scala]: Is it safe to import the global execution scheduler everywhere?

File: src/test/scala/outwatch/ScenarioTestSpec.scala [hunk 7]
```diff
@@ -37,12 +38,18 @@ class ScenarioTestSpec extends JSDomSpec {
 
     document.getElementById("minus").dispatchEvent(event)
 
-    document.getElementById("counter").innerHTML shouldBe (-1).toString
-
-    for (i <- 0 to 10) {
-      document.getElementById("plus").dispatchEvent(event)
-      document.getElementById("counter").innerHTML shouldBe i.toString
-    }
+    Task {
```
Comment (fdietze) [inline src/test/scala/outwatch/ScenarioTestSpec.scala]: This whole block looks complicated compared to the old one. Is there a simpler way to write it?

**Comment by fdietze** [comment]
@mariusmuja do you know how the `fullOpt` build size is affected?

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala (hunk 2 above)
Comment (mariusmuja) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: Without doing `startwith(Seq(Seq.empty))` and `dropWhile`, the `combinedLatest` would emit an element only when *both* the `attributeReceivers` and `childStreamReceivers` emit. This makes it emit when *either* of them emits.

File: src/main/scala/outwatch/util/Store.scala (hunk 3 above)
Comment (mariusmuja) [reply src/main/scala/outwatch/util/Store.scala]: `STRef` is used in cases where a global variable with a referential transparent API is wanted (such as the store). It's possible to remove it, but without it the store would need to be passed as an argument to all components that use it.

File: src/main/scala/outwatch/http/Http.scala (hunk 4 above)
Comment (mariusmuja) [reply src/main/scala/outwatch/http/Http.scala]: `rxscala` has an `Observable.ajax` method. Monix doesn't have that, so essentially I had to re-implement that functionality (based on `org.scalajs.dom.ext.Ajax`).

File: src/test/scala/outwatch/DomEventSpec.scala (hunk 5 above)
Comment (mariusmuja) [reply src/test/scala/outwatch/DomEventSpec.scala]: This is because Monix uses by default a [batched execution model](https://monix.io/docs/2x/execution/scheduler.html#execution-model), so we need to "fork" the assertion to the end of the javascript execution queue, otherwise it might execute before all the effects of the event dispatch execute.

File: src/main/scala/outwatch/Sink.scala (hunk 6 above)
Comment (mariusmuja) [reply src/main/scala/outwatch/Sink.scala]: It is ok, but it makes the [batched execution model](https://monix.io/docs/2x/execution/scheduler.html#execution-model) the default everywhere. It would be better to have the scheduler an implicit argument where it's needed, but that would have been a PR with more changes vs the rxscala version (which essentially doesn't have an  option of picking a different Scheduler).

File: src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 7 above)
Comment (mariusmuja) [reply src/test/scala/outwatch/ScenarioTestSpec.scala]: The simpler way would be to have the scheduler injected and use one with a synchronous execution model. Then the test code would be identical as before.

**Comment by mariusmuja** [comment]
> @mariusmuja do you know how the fullOpt build size is affected?

Because of the dead code elimination that Scala.js does, it's difficult to say precisely. But assuming that no other features of Monix are used in addition to what the library (and the tests) are using, the bundled size seems to be smaller. I did a comparison of the test bundle size a while ago and posted it in #96 , here it is:

With RxJS:
```
-rw-rw-r--   1 marius marius  313987 Nov 21 21:41 outwatch-test-opt-bundle.js.gz
-rw-rw-r--   1 marius marius 1593988 Nov 21 21:34 outwatch-test-opt-bundle.js
-rw-rw-r--   1 marius marius  837289 Nov 21 21:33 outwatch-test-opt.js
```
Monix port:
```
-rw-rw-r--   1 marius marius  240514 Nov 21 21:38 outwatch-test-opt-bundle.js.gz
-rw-rw-r--   1 marius marius 1077036 Nov 21 21:38 outwatch-test-opt-bundle.js
-rw-rw-r--   1 marius marius 1046342 Nov 21 21:37 outwatch-test-opt.js
```

File: src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 7 above)
Comment (mariusmuja) [reply src/test/scala/outwatch/ScenarioTestSpec.scala]: Actually, the simplest way would be to have the synchronous scheduler everywhere (then it would be the same execution model as rxjs and the test could have stayed the same). But I don't like that, the batched execution model is a better choice to keep the page responsive.

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala (hunk 2 above)
Comment (fdietze) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: So the behavior of combineLatest is different in monix?

Comment (mariusmuja) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: That's what I thought too, but from reading the documentation now, it seems that `combineLatest` 
 in rxjs should behave the same as in Monix. Let me look into why I had to do this (or why it's not needed with rxjs/rxscala).

File: src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 7 above)
Comment (fdietze) [reply src/test/scala/outwatch/ScenarioTestSpec.scala]: does the test fail with the batched execution model?

**Comment by fdietze** [comment]
I'm also a bit concerned that the users now have to deal with both Monix Tasks and IO. You said that they are compatible? But could outwatch be used with one of them only? How about IO and the batched execution context?

File: src/test/scala/outwatch/ScenarioTestSpec.scala (hunk 7 above)
Comment (mariusmuja) [reply src/test/scala/outwatch/ScenarioTestSpec.scala]: Actually, it was not the batched execution model that was causing the async boundary here, it was the `Observable.merge` that uses a buffered subscriber to properly implement the back-pressure protocol. I've replaced it with a simple `mergeSync`, so the test stays unchanged.

**Comment by mariusmuja** [comment]
> I'm also a bit concerned that the users now have to deal with both Monix Tasks and IO. You said that they are compatible? But could outwatch be used with one of them only? How about IO and the batched execution context?

Users don't have to deal with both. For example, `Task` is no longer used in the outwatch code (and before it was only used in the tests). 
IO doesn't know or care about the batched execution context, it's only Monix that does.

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala (hunk 2 above)
Comment (mariusmuja) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: The difference is between the  `Observable.combineLatest` rxscala companion object method and the Monix `Observable.combineLatestList`. The Monix `combineLatestList` doesn't emit an empty list if the argument list is empty, rxscala does.

**Comment by mariusmuja** [comment]
The `Scheduler` is now injected as an implicit argument where needed. For example, this allows a `TrampolineScheduler` to be injected in the two tests where we want synchronous behaviour for easier testing.

Client code will require an implicit `Scheduler` in scope, but users will get a compile error asking to import `monix.execution.Scheduler.Implicits.global` (similar to the requirement to have an  `ExecutionContext` in scope when working with futures... in fact a `Scheduler` is an `ExecutionContext`).

File: src/main/scala/outwatch/dom/helpers/DomUtils.scala (hunk 2 above)
Comment (fdietze) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: Should we have a test for this?

Comment (mariusmuja) [reply src/main/scala/outwatch/dom/helpers/DomUtils.scala]: Now the code doesn't do `.startWith(Seq(Seq.empty))....dropWhile { case (a, c) => a.isEmpty && c.isEmpty }` any more, instead it emits an `Observable(Seq.empty[Attribute])` if the `attributeReceivers` list is empty (just like in the rxscala version). There are several tests that fail if we don't do that.

File: src/main/scala/outwatch/Pipe.scala [hunk 8]
```diff
@@ -16,16 +18,16 @@ object Pipe {
     * @tparam T the type parameter of the elements
     * @return the newly created Pipe.
     */
-  def create[T](seeds: T*): IO[Pipe[T, T]] = create[T].map { pipe =>
+  def create[T](seeds: T*)(implicit s: Scheduler): IO[Pipe[T, T]] = create[T].map { pipe =>
```
Comment (LukaJCB) [inline src/main/scala/outwatch/Pipe.scala]: I wonder, since we now have both `IO` and `Task` in our project as dependencies, if we should continue using one or the other, or abstract over them with the type classes found in `cats-effect`, i.e. 
```scala
def create[F[_]: Sync, T](seeds: T*): F[Pipe[T, T]]
```
What do you guys think? :)

On an unrelated note, I also think we could get rid of the varargs here, I don't think it really ever makes sense to prepend items to a stream?

Comment (fdietze) [reply src/main/scala/outwatch/Pipe.scala]: 1. I'm in favor of abstracting over IO and Task, this would also allow using outwatch without IO Monads again.

2. One seed item is definitely enough for state-management, which is actually what we need for reactive UIs ;) (#96)

@LukaJCB could you please open PRs/Issues for this? I'll merge now.

Comment (mariusmuja) [reply src/main/scala/outwatch/Pipe.scala]: +1 on abstracting over `IO` and `Task`.

Comment (LukaJCB) [reply src/main/scala/outwatch/Pipe.scala]: I created #135 and #136 :)

File: src/main/scala/outwatch/Sink.scala [hunk 9]
```diff
@@ -48,35 +50,39 @@ sealed trait Sink[-T] extends Any {
...
-    override private[outwatch] def observer = sink.observer
+  private[outwatch] final case class ObservableSink[-I, +O](oldSink: Sink[I], stream: Observable[O]) extends Observable[O] with Sink[I] {
+    override private[outwatch] def observer = oldSink.observer
+
+    override def unsafeSubscribeFn(subscriber: Subscriber[O]): Cancelable = stream.unsafeSubscribeFn(subscriber)
   }
 
-  private[outwatch] final case class SubjectSink[T]() extends Subject[T](new SubjectFacade) with Sink[T] {
-    override private[outwatch] def observer = this
+  private[outwatch] final case class SubjectSink[T]()(implicit scheduler: Scheduler) extends Observable[T] with Sink[T] {
```
Comment (LukaJCB) [inline src/main/scala/outwatch/Sink.scala]: Why exactly do we need the scheduler here? Could we not add it to the methods? This would unclutter our code a bit, no?

Comment (LukaJCB) [reply src/main/scala/outwatch/Sink.scala]: In fact, why don't we do:

`def observer: Observer[T]`? So we wouldn't need the scheduler at all?

Comment (mariusmuja) [reply src/main/scala/outwatch/Sink.scala]: We need it to create the `Subscriber` inside the `SubjectSink` (the `Sink.observer` is actually a `Subscriber` now) . If I would have left it an `Observer`, all the `Sink.redirect` operations would have required a `Scheduler` (and also many operations in `PipeOps`). This way they just use the `Scheduler` from the source `Sink`'s `Subscriber`.

Comment (LukaJCB) [reply src/main/scala/outwatch/Sink.scala]: Yeah, that makes sense, however, I think creating a new `Handler` or `Pipe` is going to be a much much more common operation, then `redirect`, so I think it'd be better if the individual methods required the implicit evidence, no?

Comment (mariusmuja) [reply src/main/scala/outwatch/Sink.scala]: Not quite, `EmitterBuilder`'s `-->` would require an implicit `Scheduler` in that case.
